clippy_lints/src/methods/iter_on_single_or_empty_collections.rs

```diff
@@ -43,6 +65,25 @@ pub(super) fn check(cx: &LateContext<'_>, expr: &Expr<'_>, method_name: &str, re
     let is_unified = match get_expr_use_or_unification_node(cx.tcx, expr) {
         Some((Node::Expr(parent), child_id)) => match parent.kind {
             ExprKind::If(e, _, _) | ExprKind::Match(e, _, _) if e.hir_id == child_id => false,
+            ExprKind::Call(
+                Expr {
+                    kind: ExprKind::Path(path),
+                    hir_id,
+                    ..
+                },
+                args,
+            ) => cx
+                .typeck_results()
+                .qpath_res(path, *hir_id)
+                .opt_def_id()
+                .filter(|fn_id| cx.tcx.def_kind(fn_id).is_fn_like())
+                .is_some_and(|fn_id| is_arg_ty_unified_in_fn(cx, fn_id, expr.hir_id, args)),
```

Comment:
`expr` here is not always necessarily a function argument because `get_expr_use_or_unification_node` skips certain kinds of expressions such as blocks, so the assumption that you'll always find the `.iter()` call expression in the list of fn arguments seems wrong and the unwrap above could panic.

Can you add a test case where the function argument is `{ [].iter() }` (with a block specifically)?

Reply:
Did that, after a small adjustment `{ [].iter() }` is also fine

Reply:
Normally I'd be more comfortable with avoiding unwrap and getting a false negative in the worst case but that logic looks correct now so it should be fine.